src/emulator/cpu.ts

```diff
@@ -109,26 +109,76 @@ export class Cpu {
         this.interrupt.clear(irq);
         this.state.interruptsEnabled = false;
 
-        this.state.r16[r16.sp] = (this.state.r16[r16.sp] - 1) & 0xffff;
-        this.bus.write(this.state.r16[r16.sp], this.state.p >>> 8);
-        this.state.r16[r16.sp] = (this.state.r16[r16.sp] - 1) & 0xffff;
-        this.bus.write(this.state.r16[r16.sp], this.state.p & 0xff);
+        this.stackPush(this.state.p);
 
         this.state.p = getIrqVector(irq);
         this.clock.increment(5);
 
         return 5;
     }
 
+    private stackPush(value: number): void {
+        value = value & 0xffff;
+        this.state.r16[r16.sp] = (this.state.r16[r16.sp] - 0x01) & 0xffff;
+        this.bus.write(this.state.r16[r16.sp], value >>> 8);
+        this.state.r16[r16.sp] = (this.state.r16[r16.sp] - 0x01) & 0xffff;
+        this.bus.write(this.state.r16[r16.sp], value & 0xff);
+    }
+
+    private stackPop(): number {
+        const value = this.bus.read16(this.state.r16[r16.sp]) & 0xffff;
+        this.state.r16[r16.sp] = (this.state.r16[r16.sp] + 2) & 0xffff;
+
+        return value;
+    }
+
     private dispatch(instruction: Instruction): number {
         if (instruction.op !== Operation.invalid) this.onExecute.dispatch(this.state.p);
 
         switch (instruction.op) {
+            case Operation.adc: {
+                this.clock.increment(instruction.cycles);
+
+                const a = this.state.r8[r8.a];
+                const operand = this.getArg1(instruction);
+                const result = a + operand + (this.state.r8[r8.f] & flag.c);
```

Comment:
I am pretty sure this should be  `(this.state.r8[r8.f] & flag.c) >>> 4` --- carry is bit 4. This is usually used for 16 bit additions by first adding the low bytes and then adding the high bytes with carry, so an overflow on the low bytes amounts to `carry high` amounts to an additional contribution of 1 to the high bytes.